src/_pytest/python.py

```diff
@@ -1220,32 +1278,41 @@ def _idvalset(idx, parameterset, argnames, idfn, ids, item, config):
         ]
         return "-".join(this_id)
     else:
-        return _ascii_escaped_by_config(ids[idx], config)
-
-
-def idmaker(argnames, parametersets, idfn=None, ids=None, config=None, item=None):
-    ids = [
+        id = ids[idx]
+        assert id is not None  # Checked above.
+        return _ascii_escaped_by_config(id, config)
+
+
+def idmaker(
+    argnames: Iterable[str],
+    parametersets: Iterable[ParameterSet],
+    idfn: Optional[Callable[[object], Optional[object]]] = None,
+    ids: Optional[List[Union[None, str]]] = None,
+    config: Optional[Config] = None,
+    item=None,
+) -> List[str]:
+    resolved_ids = [
```

Comment:
Renamed so can have a different type from the input `ids`.

Reply:
Maybe s/resolved_ids/ret/ then?

Reply:
I used this name mostly because of the terminology established by `_resolve_arg_ids`. Can change if you think `ret` is better.

Reply:
It's fine then.